### DISCUSSION TITLE
How to inherit for Profiles & Certificates (and probably Instructions, Extensions, etc.)

### DISCUSSION THREAD

**Discussion by james-ball-qualcomm** [riscv-software-src/riscv-unified-db#247] [General]
@dhower-qc has proposed using $mref and I believe it has been merged into main. However, @drom raised concerns about it today.

Could we have a generalized facility in UDB that when an object of a particular kind is created, if another compatible object is provided to inherit from it could do this automatically for UDB objects we care about. I'm thinking of something that could copy the @data from the inherit object and then merge it with the YAML for the new object being created. This is pretty hand-wavy but hopefully it is a useful suggestion?

**Comment by dhower-qc** [comment]
+@drom

BTW, there is a merge facility built-in to YAML. Thus far we've tried to keep the YAML bare-bones (strict superset of JSON), but it's worth considering:

https://yaml.org/type/merge.html

**Comment by dhower-qc** [comment]
Came across this. (a subset of) the operators look about right.

https://www.npmjs.com/package/json-merger

**Comment by james-ball-qualcomm** [comment]
I've added support for the $inherit YAML extension that Derek has created (formerly called $mref). It is working well for me in all the profiles and certificate use cases. So, I think we can close this discussion.

**Comment by drom** [comment]
It might be worth looking at the immutable update rather then inheritance model.
Here is set of immutable methods we can use: https://github.com/kolodny/immutability-helper?tab=readme-ov-file#available-commands

**Comment by james-ball-qualcomm** [comment]
Maybe we could use a merge as long as it was deep?

+james+
________________________________

**Comment by dhower-qc** [comment]
How would immutable update be used in the data files? From what I understand, it's more of a post-processing command than something to represent data in the source format.
